File: Recursion-Backtracking/07-Subset-Sum-II/subset-sum-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

function findSubsets(index, n, curr, result, m, nums){
    if(index === n){
        const temp = [...curr].sort((a,b) => a - b);
        const sortkey = JSON.stringify(temp);
        if(!m.hasOwnProperty(sortkey)){
            result.push(temp);
            m[sortkey] = true;
        }
        return;
    }
    findSubsets(index + 1, n, curr, result, m, nums);
    curr.push(nums[index]);
    findSubsets(index + 1, n, curr, result, m, nums);
    curr.pop();
}

var subsetsWithDup = function(nums) {
    let result = [];
    let curr = [];
    let m = {};

    findSubsets(0, nums.length, curr, result, m, nums);

    return result;
};

File: Recursion-Backtracking/07-Subset-Sum-II/subset-sum-ii.cpp
class Solution {
public:

    void findSubsets(int index, int n, vector<int>& nums, vector<int>& curr, set<vector<int>>& s){
        if(index == n){
            vector<int> temp(curr.begin(), curr.end());
            sort(temp.begin(), temp.end());
            s.insert(temp);
            return;
        }

        findSubsets(index + 1, n, nums, curr, s);
        curr.push_back(nums[index]);
        findSubsets(index + 1, n, nums, curr, s);
        curr.pop_back();
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        set<vector<int>> s;
        vector<int> curr;

        findSubsets(0, nums.size(), nums, curr, s);

        vector<vector<int>> result(s.begin(), s.end());
        return result;
    }
};